source/s3_auto_ranged_get.c

```diff
@@ -117,6 +119,16 @@ static enum aws_s3_auto_ranged_get_request_type s_s3_get_request_type_for_discov
     struct aws_s3_auto_ranged_get *auto_ranged_get = meta_request->impl;
     AWS_ASSERT(auto_ranged_get);
 
+    /*
+     * If the object is empty, download the file using a GetObject with PartNumber request. This will succeed if
+     * the file is empty, or it will download the file.
+     */
```

Comment:
this explains what we're doing, but there are subtle reasons we're doing it this way that should be explained somewhere 

explain why we're bothering to send another request, when we already sent one and there's no body anyway

and explain why we're downloading with part-number, instead of something simpler like a basic GetObject

If this isn't the right place for that explanation, tell the reader here where they can find that explanation

Reply:
`HeadObject` doesn't fail, and we do get the Etag, and even the checksum if it's available. After `HeadObject`, if the file is empty, we just complete the request. I have added a test for this case [here] (https://github.com/awslabs/aws-c-s3/pull/389/files#diff-0f16837166c257c5841490cb42e6840835b93fbbcdc798e896ec0728bbe07bd5R3604).

Reply:
I have added more details to the documentation about the flow. 

`HeadObject` doesn't fail, and we do get the ETag, and even the checksum, if available. After `HeadObject`, if the file is empty, we simply complete the request. I have added a test for this case and fixed a bug. Now we have the following `empty_file` cases; let me know if I have missed any:
- We execute `AWS_S3_AUTO_RANGE_GET_REQUEST_TYPE_GET_OBJECT_WITH_RANGE`; it fails, and we download the file with `AWS_S3_AUTO_RANGE_GET_REQUEST_TYPE_GET_OBJECT_WITH_PART_NUMBER_1`.
- We execute `AWS_S3_AUTO_RANGE_GET_REQUEST_TYPE_GET_OBJECT_WITH_PART_NUMBER_1`, and it succeeds with a checksum.
- We perform `HeadObject`, and it succeeds with a checksum.
- We execute `AWS_S3_AUTO_RANGE_GET_REQUEST_TYPE_GET_OBJECT_WITH_RANGE`, and the user has provided a range header; it fails with a 416 RANGE_NOT_SATISFIABLE error.